### PR TITLE
VXQUERY-209, 207 - Parallel index creation and access and index centralization

### PR DESCRIPTION
**Pull request by menakaj**
This pull request contains fixes for several indexing issues.
1. VXQuery 207/ 197 - Index Centralization
2. VXQuery 209 - Parallel index creation and access
3. VXQuery 220 - Show index query

And fixing following bug.
VXQuery 235 - Exception when executing Aggregation queries.
[](https://issues.apache.org/jira/browse/VXQUERY-235)

This also includes the indexing user documentation.

Please review.

### CODE REVIEW

File: src/site/apt/user_indexing.apt [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+~~ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+~~ See the License for the specific language governing permissions and
+~~ limitations under the License.
+
+How to use Indexing Features in VXQuery.
+
+In VXQuery, all the indexes are created in user specified directory. So, before you start using indexing functions,
+you should do the following.
+
+*** Configuring VXQuery to use indexing functions.
```
Comment (prestoncarman) [inline src/site/apt/user_indexing.apt:21]: No coding changes should be required for the user. I think we should point them to the cluster xml configuration file to specify the index location. I think setting up a cluster is a prerequisite for using indexing.

File: vxquery-core/src/test/java/org/apache/vxquery/indexing/MetaFileUtilTest.java [hunk 2]
```diff
@@ -49,7 +49,8 @@
     @BeforeClass
     public static void setup() {
         new File(TestConstants.INDEX_DIR).mkdir();
-        metaFileUtil = MetaFileUtil.create(TestConstants.INDEX_DIR);
+        //        metaFileUtil = MetaFileUtil.create(TestConstants.INDEX_DIR);
```
Comment (prestoncarman) [inline vxquery-core/src/test/java/org/apache/vxquery/indexing/MetaFileUtilTest.java:52]: Remove commented out code.

File: vxquery-core/src/main/java/org/apache/vxquery/compiler/rewriter/rules/PushChildIntoDataScanRule.java [hunk 3]
```diff
@@ -69,6 +70,8 @@
 
     protected boolean processOperator(Mutable<ILogicalOperator> opRef, IOptimizationContext context)
             throws AlgebricksException {
+
```
Comment (prestoncarman) [inline vxquery-core/src/main/java/org/apache/vxquery/compiler/rewriter/rules/PushChildIntoDataScanRule.java:73]: Formatting check.

File: vxquery-core/src/main/java/org/apache/vxquery/compiler/rewriter/rules/AbstractCollectionRule.java [hunk 4]
```diff
@@ -78,11 +83,9 @@
...
             return null;
         }
-        AbstractFunctionCallExpression functionCall = (AbstractFunctionCallExpression) logicalExpression;
-        if (!functionCall.getFunctionIdentifier()
-                .equals(BuiltinFunctions.FN_COLLECTION_WITH_TAG_2.getFunctionIdentifier())
-                && !functionCall.getFunctionIdentifier()
-                        .equals(BuiltinFunctions.FN_COLLECTION_1.getFunctionIdentifier())) {
+        functionCall = (AbstractFunctionCallExpression) logicalExpression;
+
+        if (!functions.contains(functionCall.getFunctionIdentifier())){
```
Comment (prestoncarman) [inline vxquery-core/src/main/java/org/apache/vxquery/compiler/rewriter/rules/AbstractCollectionRule.java:88]: Formatting Check.

File: vxquery-core/src/main/java/org/apache/vxquery/metadata/VXQueryMetadataProvider.java [hunk 5]
```diff
@@ -88,22 +88,43 @@ public StaticContext getStaticContext() {
             List<LogicalVariable> minFilterVars, List<LogicalVariable> maxFilterVars, IOperatorSchema opSchema,
             IVariableTypeEnvironment typeEnv, JobGenContext context, JobSpecification jobSpec, Object implConfig)
                     throws AlgebricksException {
-        VXQueryCollectionDataSource ds = (VXQueryCollectionDataSource) dataSource;
+        VXQueryCollectionDataSource ds = null;
+        VXQueryIndexingDataSource ids = null;
+
+        try {
+            ids = (VXQueryIndexingDataSource) dataSource;
+        } catch (ClassCastException e){
```
Comment (prestoncarman) [inline vxquery-core/src/main/java/org/apache/vxquery/metadata/VXQueryMetadataProvider.java:96]: Formating check

File: vxquery-core/src/main/java/org/apache/vxquery/runtime/functions/index/indexCentralizer/IndexCentralizerUtil.java [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+    private final String FILE_NAME = "VXQuery-Index-Directory.xml";
+    private final List<String> collections = new ArrayList<>();
+    private final Logger LOGGER = Logger.getLogger("IndexCentralizerUtil");
+    private File XML_FILE;
+    private String INDEX_LOCATION;
+    private ConcurrentHashMap<String, IndexLocator> indexCollectionMap = new ConcurrentHashMap<>();
+
+    public IndexCentralizerUtil(File index) {
+        this.INDEX_LOCATION = index.getPath();
+        if (!index.exists()){
```
Comment (prestoncarman) [inline vxquery-core/src/main/java/org/apache/vxquery/runtime/functions/index/indexCentralizer/IndexCentralizerUtil.java:60]: Formatting check

**Comment by prestoncarman** [comment]
Do we have tests for parallel indexing (for 1, 2 and 4 partitions)?

**Comment by prestoncarman** [comment]
I think any major issue with this PR can be fixed later. Lets just create issues for any of these issues.
For example:
Can you create an issue for updating indexing to support the "source" variable (as I understand it uses absolute paths instead of relative)?

**Comment by menakaj** [comment]
Yes it uses absolute path. I'll create a Jira issue for it.

File: src/site/apt/user_indexing.apt (hunk 1 above)
Comment (menakaj) [reply src/site/apt/user_indexing.apt:21]: But, it is being read only when starting the vxquery server. How could someone use this without a cluster?

Comment (prestoncarman) [reply src/site/apt/user_indexing.apt:21]: If we want to support a user to use an index outside of a cluster, then I think having an additional cli argument would be a good option.

**Comment by prestoncarman** [comment]
This code has been merged. PR can be closed.

**Pull request closed without merging**